File: src/main/java/com/derongan/minecraft/madeinabyss/Ascension/Effect/AbstractAscensionEffect.java
package com.derongan.minecraft.madeinabyss.Ascension.Effect;

import com.derongan.minecraft.madeinabyss.AbyssContext;
import org.bukkit.entity.Player;

public abstract class AbstractAscensionEffect implements AscensionEffect {
    private AbyssContext context;
    int duration;
    int strength;
    private long offset; //TODO honor this value

    AbstractAscensionEffect(AbyssContext context, long offset, int strength, int duration) {
        this.context = context;
        this.duration = duration;
        this.offset = offset;
        this.strength = strength;
    }

    AbyssContext getContext() {
        return context;
    }

    @Override
    public void applyEffect(Player player, int ticks) {
        duration -= ticks;
        applyEffect(player);
    }

    abstract void applyEffect(Player player);

    @Override
    public boolean isDone() {
        return duration <= 0;
    }

    @Override
    public int getRemainingTicks() {
        return duration;
    }

    @Override
    public void setRemainingTicks(int remainingTicks) {
        duration = remainingTicks;
    }

    @Override
    public void cleanUp(Player player) {
    }
}

File: src/main/java/com/derongan/minecraft/mineinabyss/ascension/effect/AbstractAscensionEffect.kt
package com.derongan.minecraft.mineinabyss.ascension.effect

import com.derongan.minecraft.mineinabyss.mineInAbyss
import com.mineinabyss.idofront.time.TimeSpan
import com.okkero.skedule.schedule
import org.bukkit.Bukkit
import org.bukkit.entity.Player

abstract class AbstractAscensionEffect : AscensionEffect {
    abstract val offset: TimeSpan
    abstract val iterations: Int

    override fun applyEffect(player: Player, ticks: Int) {
        mineInAbyss.schedule {
            waitFor(offset.ticks)
            repeat(iterations) {
                applyEffect(player)
                waitFor(duration.ticks)
            }
            cleanUp(Bukkit.getPlayer(player.uniqueId) ?: player) //get new player reference if player relogged
        }
    }

    abstract fun applyEffect(player: Player)

    override val isDone: Boolean
        get() = duration.millis <= 0

    override fun cleanUp(player: Player) {}
}